Training/interface/DataLoader_main.h

```diff
@@ -0,0 +1,770 @@
+        // trainingWeightFactor(tauTuple.GetEntries() / training_weight_factor,
+        innerCellGridRef(n_inner_cells, n_inner_cells, inner_cell_size, inner_cell_size),
+        outerCellGridRef(n_outer_cells, n_outer_cells, outer_cell_size, outer_cell_size)
+    {
+      if(n_threads > 1) ROOT::EnableImplicitMT(n_threads);
+    }
+
+    void Initialize(std::string file_name)
+    {
+        if(tauTuple)
+            throw std::runtime_error("DataLoader is already initialized.");
+        file = OpenRootFile(file_name);
+        tauTuple = std::make_shared<tau_tuple::TauTuple>(file.get(), true);
+    }
+
+    bool HasNext() {
+        return (current_entry + n_tau) < end_dataset;
+    }
+
+    std::shared_ptr<Data> LoadNext(){
+
+        if(!tauTuple)
+            throw std::runtime_error("DataLoader is not initialized.");
+
+        auto data = std::make_shared<Data>(n_tau, n_fe_tau,
+                                           n_inner_cells, n_outer_cells,
+                                           n_pf_el,
+                                           n_pf_mu,
+                                           n_pf_chHad,
+                                           n_pf_nHad,
+                                           n_pf_gamma,
+                                           n_electrons,
+                                           n_muons,
+                                           n_labels
+                                          );
+
+        const Long64_t end_entry = current_entry + n_tau;
+        size_t n_processed = 0, n_total = static_cast<size_t>(end_entry - start_dataset);
+        for(Long64_t tau_i = 0; tau_i < n_tau; ++tau_i, ++current_entry) {
+            tauTuple->GetEntry(current_entry);
+            const auto& tau = tauTuple->data();
+            if(parity == -1 || tau.evt % 2 == static_cast<unsigned>(parity)) {
+                FillTauBranches(tau, tau_i, data);
+                FillCellGrid(tau, tau_i, innerCellGridRef, data, true);
+                FillCellGrid(tau, tau_i, outerCellGridRef, data, false);
+            }
+        }
+
+        std::cout << "One batch is returned" << std::endl;
+        return data;
+    }
+
+  private:
+      static constexpr float pi = boost::math::constants::pi<float>();
+
+      template<typename Scalar>
+      static Scalar DeltaPhi(Scalar phi1, Scalar phi2)
+      {
+          static constexpr Scalar pi = boost::math::constants::pi<Scalar>();
+          Scalar dphi = phi1 - phi2;
+          if(dphi > pi)
+              dphi -= 2*pi;
+          else if(dphi <= -pi)
+              dphi += 2*pi;
+          return dphi;
+      }
+
+
+      void FillTauBranches(const Tau& tau, Long64_t tau_i, std::shared_ptr<Data>& data)
+      {
+        Long64_t start_array_index = tau_i * n_fe_tau;
+
+        // Filling Tau Branche
+        auto get_tau_branch = [&](TauFlat_f _fe) -> float& {
+            size_t _fe_ind = static_cast<size_t>(_fe);
```

Comment:
here and in the similar lambdas below: add check that _fe >=0. Setting some variables to -1, it would provide as with an easy way of disabling them.

Reply:
I see only the following way to do it:

auto fill_PFelectron = [&](PfCand_electron_f _fe_n, float value) {
size_t _idx = static_cast<size_t>(_fe_n);
if(_idx >= 0){
size_t flat_index = getIndex(n_pf_el, _idx);
return data->x_grid.at(CellObjectType::PfCand_electron).at(inner).at(flat_index);
}
};

and that fill branches like this: fill_PFelectron(Br::pfCand_ele_rel_pt, valid ? tau.pfCand_pt.at(pfCand_idx) / tau.tau_pt : 0);
would you say that it is fine to change everything like this? otherwise I have to allocate additional pointer and return reference to it, I do not see other options.

Reply:
Why something like this wouldn't work?:
```c++
float foo;

auto get_tau_branch = [&](TauFlat_f _fe) -> float& {
    if(_fe < 0) return foo;
    size_t _fe_ind = static_cast<size_t>(_fe);
    size_t index = start_array_index + _fe_ind;
    return data->x_tau.at(index);
};
```

similarly for the other cases.